utils/clickhouse.h

```diff
@@ -0,0 +1,134 @@
+/**
+ *		Tempesta FW
+ *
+ * Copyright (C) 2024 Tempesta Technologies, Inc.
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License,
+ * or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along with
+ * this program; if not, write to the Free Software Foundation, Inc., 59
+ * Temple Place - Suite 330, Boston, MA 02111-1307, USA.
+ */
+
+#ifndef __TFW_CLICKHOUSE_H__
+#define __TFW_CLICKHOUSE_H__
+
+#include <iostream>
+#include <clickhouse/client.h>
+#include <clickhouse/base/socket.h>
+
+using namespace clickhouse;
+using namespace std;
+
+/**
+ * Class for sending records to a Clickhouse database.
+ * It manages multiple instances of the Clickhouse client (one per thread) to
+ * handle concurrent operations, as the Clickhouse client itself is not
+ * thread-safe.
+ *
+ * Constructor:
+ *    @TfwClickhouse - Initializes the Clickhouse connection for each CPU core,
+ *        setting up individual clients and data blocks, with a provided callback.
+ *
+ * Destructor:
+ *    @~TfwClickhouse - Cleans up by deleting all Client and Block instances
+ *        and freeing the tasks array.
+ *    @getBlock - Returns a pointer to the data block for the specified CPU core.
+ *    @commit - Commits the data in the block to the Clickhouse database if the
+ *        elapsed time since the last insertion exceeds a predefined threshold
+ *        or the block’s row count exceeds a maximum event threshold. After
+ *        committing, the block is deleted, a new block is created via
+ *        block_callback and last_time is updated.
+ *
+ * Private Struct:
+ *    @struct TfwTask - Holds task-specific data for each CPU core, including:
+ *        @client - Clickhouse Client instance for sending data to the database.
+ *        @block - Block instance holding data records to be inserted.
+ *        @last_time - The last timestamp when data was sent.
+ *
+ * Private Members:
+ *    @tasks - Array of tasks, one for each CPU core, to manage per-thread data
+ *        and clients.
+ *    @block_callback - Callback function that creates a new Block for data
+ *        storage.
+ *    @cpu_cnt - Number of CPU cores to be used.
+ *    @table_name - Name of the Clickhouse table where data is inserted.
+ */
+class TfwClickhouse {
+public:
+	TfwClickhouse(string host, string table_name,
+		      unsigned int cpu_cnt, Block *(*cb)());
+	~TfwClickhouse();
+	Block *getBlock(unsigned int cpu);
```

Comment:
It seems we should implement copy and move constructors for this class. May be make them private to Prohibit copy

Reply:
Made it in such a way:
```
	TfwMmapBufferReader(const TfwMmapBufferReader &) = delete;
	TfwMmapBufferReader &operator=(const TfwMmapBufferReader &) = delete;
```